src/components/structures/MatrixChat.tsx

```diff
@@ -1588,152 +1604,168 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
-            };
-
-            // on our URLs there might be a ?via=matrix.org or similar to help
-            // joins to the room succeed. We'll pass these through as an array
-            // to other levels. If there's just one ?via= then params.via is a
-            // single string. If someone does something like ?via=one.com&via=two.com
-            // then params.via is an array of strings.
-            let via = [];
-            if (params.via) {
-                if (typeof(params.via) === 'string') via = [params.via];
-                else via = params.via;
-            }
+                break;
+            case Screens.FORGOT_PASSWORD:
+                dis.dispatch({
+                    action: 'start_password_recovery',
+                    params: params,
+                });
+                break;
+            case Screens.SOFT_LOGOUT:
+                if (MatrixClientPeg.get() && MatrixClientPeg.get().getUserId() && !Lifecycle.isSoftLogout()) {
+                    // Logged in - visit a room
+                    this.viewLastRoom();
+                } else {
+                    // Ultimately triggers soft_logout if needed
+                    dis.dispatch({
+                        action: 'start_login',
+                        params: params,
+                    });
+                }
+                break;
+            case Screens.NEW:
+                dis.dispatch({
+                    action: 'view_create_room',
+                });
+                break;
+            case Screens.SETTINGS:
+                dis.dispatch({
+                    action: 'view_user_settings',
+                });
+                break;
+            case Screens.WELCOME:
+                dis.dispatch({
+                    action: 'view_welcome_page',
+                });
+                break;
+            case Screens.HOME:
+                dis.dispatch({
+                    action: 'view_home_page',
+                });
+                break;
+            case Screens.START:
+                this.showScreen(Screens.HOME);
+                dis.dispatch({
+                    action: 'require_registration',
+                });
+                break;
+            case Screens.DIRECTORY:
+                dis.dispatch({
+                    action: 'view_room_directory',
+                });
+                break;
+            case Screens.GROUPS:
+                dis.dispatch({
+                    action: 'view_my_groups',
+                });
+                break;
+            case Screens.COMPLETE_SECURITY:
+                dis.dispatch({
+                    action: 'start_complete_security',
+                });
+                break;
+            case Screens.POST_REGISTRATION:
+                dis.dispatch({
+                    action: 'start_post_registration',
+                });
+                break;
+            default:
+                if (screen.startsWith('room/')) {
```

Comment:
hmm... didn't realize this was here. Kinda made sense in being an if-else ladder before ._.

Reply:
ok. reverting.